### PR TITLE
properly handle uncompressed chunk leak when using revrange

### PR DESCRIPTION
**Pull request by danni-m**

### CODE REVIEW

File: src/compressed_chunk.c [hunk 1]
```diff
@@ -188,13 +188,14 @@ u_int64_t getIterIdx(ChunkIter_t *iter) {
...
 // LCOV_EXCL_STOP
 
-ChunkIter_t *Compressed_NewChunkIterator(Chunk_t *chunk, bool rev) {
+ChunkIter_t *Compressed_NewChunkIterator(Chunk_t *chunk, int options) {
     CompressedChunk *compressedChunk = chunk;
 
     // for reverse iterator of compressed chunks
-    if (rev == true) {
+    if (options & CHUNK_ITER_OP_REVERSE) {
+        int uncompressed_options = CHUNK_ITER_OP_REVERSE | CHUNK_ITER_OP_FREE_CHUNK;
```
Comment (gkorland) [inline src/compressed_chunk.c:196]: @danni-m why do you need to define `uncompressed_options` with `CHUNK_ITER_OP_FREE_CHUNK`? I don't see that this flag is tested in `Uncompressed_NewChunkIterator`.

Why not just?
```
         return Uncompressed_NewChunkIterator(uncompressedChunk, CHUNK_ITER_OP_REVERSE); 
```

File: src/generic_chunk.h [hunk 2]
```diff
@@ -23,6 +23,10 @@ typedef struct Sample {
 typedef void Chunk_t;
 typedef void ChunkIter_t;
 
+#define CHUNK_ITER_OP_REVERSE 1
```
Comment (ashtul) [inline src/generic_chunk.h:26]:
```suggestion
#define CHUNK_ITER_OP_NONE 0
#define CHUNK_ITER_OP_REVERSE 1
```

File: src/tsdb.c [hunk 3]
```diff
@@ -616,14 +625,14 @@ timestamp_t getFirstValidTimestamp(Series *series, long long *skipped) {
     RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(series->chunks, "^", NULL, 0);
     RedisModule_DictNextC(iter, NULL, (void *)&chunk);
 
-    ChunkIter_t *chunkIter = funcs->NewChunkIterator(chunk, rev);
+    ChunkIter_t *chunkIter = funcs->NewChunkIterator(chunk, 0);
```
Comment (ashtul) [inline src/tsdb.c:628]: use `CHUNK_ITER_OP_NONE` instead of 0 for all `NewChunkIterator` functions

File: src/compressed_chunk.c [hunk 4]
```diff
@@ -217,11 +218,7 @@ ChunkResult Compressed_ChunkIteratorGetNext(ChunkIter_t *iter, Sample *sample) {
     return Compressed_ReadNext((Compressed_Iterator *)iter, &sample->timestamp, &sample->value);
 }
 
-void Compressed_FreeChunkIterator(ChunkIter_t *iter, bool rev) {
-    // compressed iterator on reverse query has to release decompressed chunk
-    if (rev) {
-        free(((ChunkIterator *)iter)->chunk);
-    }
+void Compressed_FreeChunkIterator(ChunkIter_t *iter) {
```
Comment (gkorland) [inline src/compressed_chunk.c:221]: @danni-m don't you need to free the chunk also in the compressed case?

```
    if (iter->options & CHUNK_ITER_OP_FREE_CHUNK) {
        free(iter->chunk);
    }
```

File: src/chunk.c [hunk 5]
```diff
@@ -144,13 +144,13 @@ ChunkResult Uncompressed_UpsertSample(UpsertCtx *uCtx, int *size) {
...
 }
 
-ChunkIter_t *Uncompressed_NewChunkIterator(Chunk_t *chunk, bool rev) {
+ChunkIter_t *Uncompressed_NewChunkIterator(Chunk_t *chunk, int options) {
     ChunkIterator *iter = (ChunkIterator *)calloc(1, sizeof(ChunkIterator));
     iter->chunk = chunk;
-    if (rev == false) { // iterate from first to last
-        iter->currentIndex = 0;
-    } else { // iterate from last to first
+    if (options & CHUNK_ITER_OP_REVERSE) { // iterate from last to first
```
Comment (ashtul) [inline src/chunk.c:150]:
```suggestion
    iter->currentIndex = (options & CHUNK_ITER_OP_REVERSE) ? iter->chunk->num_samples - 1 : 0;
```

File: src/chunk.c [hunk 6]
```diff
@@ -177,8 +177,11 @@ ChunkResult Uncompressed_ChunkIteratorGetPrev(ChunkIter_t *iterator, Sample *sam
     }
 }
 
-void Uncompressed_FreeChunkIterator(ChunkIter_t *iter, bool rev) {
-    (void)rev; // only used with compressed chunk but signature must be similar
+void Uncompressed_FreeChunkIterator(ChunkIter_t *iterator) {
+    ChunkIterator *iter = (ChunkIterator *)iterator;
+    if (iter->options & CHUNK_ITER_OP_FREE_CHUNK) {
```
Comment (ashtul) [inline src/chunk.c:182]: Great encapsulation!

**Review by ashtul — COMMENTED** [review]
Looks good.
Keeping the `free_chunk` flag is better than relying on the rev.
Small style comments

File: src/compressed_chunk.c (hunk 1 above)
Comment (ashtul) [reply src/compressed_chunk.c:196]: @gkorland 
B/c some in regular range we don't free as we don't own the chunks
In revenge, we decompressed and therefore we need to free

Comment (danni-m) [reply src/compressed_chunk.c:196]: To expand what ariel wrote, an iterator "owns" a chunk (and not the other way around) only when a compressed chunk is iterated in reverse. So logically we should only support that specific scenario as the compressed chunk iterator should never own the actual chunk.

File: src/compressed_chunk.c (hunk 4 above)
Comment (danni-m) [reply src/compressed_chunk.c:221]: see comment above.

File: src/chunk.c (hunk 5 above)
Comment (danni-m) [reply src/chunk.c:150]: I generally avoid one-liners ifs as they are more complex to read and comprehend when reading complex code bases.

File: src/compressed_chunk.c [hunk 7]
```diff
@@ -106,7 +106,7 @@ ChunkResult Compressed_UpsertSample(UpsertCtx *uCtx, int *size) {
     size_t newSize = oldChunk->size;
 
     CompressedChunk *newChunk = Compressed_NewChunk(newSize);
-    Compressed_Iterator *iter = Compressed_NewChunkIterator(oldChunk, false);
+    Compressed_Iterator *iter = Compressed_NewChunkIterator(oldChunk, 0);
```
Comment (ashtul) [inline src/compressed_chunk.c:109]: CHUNK_ITER_OP_NONE

File: src/compressed_chunk.c [hunk 8]
```diff
@@ -75,7 +75,7 @@ Chunk_t *Compressed_SplitChunk(Chunk_t *chunk) {
     // add samples in new chunks
     size_t i = 0;
     Sample sample;
-    ChunkIter_t *iter = Compressed_NewChunkIterator(curChunk, false);
+    ChunkIter_t *iter = Compressed_NewChunkIterator(curChunk, 0);
```
Comment (ashtul) [inline src/compressed_chunk.c:78]: CHUNK_ITER_OP_NONE

**Review by ashtul — COMMENTED** [review]
these  two as well

**Review by ashtul — APPROVED** [review]

**Pull request merged**